File: samples/cpp/security.cpp
/*
 *  security.cpp — the security layers observed from client code.
 *
 *  Companion to ../../security-architecture.md.  In one run:
 *    1. WHO AM I    — the attachment's own MON$ATTACHMENTS row: auth plugin
 *                     (Srp256) and wire-crypt plugin (ChaCha64), i.e. layers
 *                     1 and 2 as the server recorded them.
 *    2. USERS       — SEC$USERS, the virtual view over the security database.
 *    3. LEAST PRIVILEGE — a temporary user plus a role carrying the
 *                     MONITOR_ANY_ATTACHMENT system privilege; connecting
 *                     with and without the role changes how much of
 *                     MON$ATTACHMENTS the same user can see.
 *    4. FAILED LOGIN — the error chain a wrong password produces.
 *  All DDL happens on a scratch database / the security database; the demo
 *  cleans up its user and role at the end.
 */

#include "fb_sample.h"

using namespace Firebird;

static const char* DSN = "inet://localhost//tmp/fbhandson/security.fdb";
static const char* TMP_USER = "HANDSON_USER";
static const char* TMP_PASS = "Hands0nPw";

// Run one statement in its own transaction, ignoring any error — needed for
// idempotent cleanup, because DROP USER reports "record not found" only at
// COMMIT (user management is deferred work, executed at commit time).
static void execIgnore(fbsample::Db& db, const char* sql)
{
	ITransaction* t = nullptr;
	try
	{
		t = db.start();
		db.exec(t, sql);
		t->commit(&db.status);
	}
	catch (const FbException&)
	{
		if (t)
		{
			CheckStatusWrapper quiet(fbsample::master->getStatus());
			t->rollback(&quiet);
			quiet.dispose();
		}
	}
}

static void whoAmI(fbsample::Db& db, ITransaction* tra, const char* label)
{
	fbsample::Db::Table t = db.query(tra,
		"select trim(mon$user), mon$auth_method, mon$wire_crypt_plugin, mon$remote_protocol,"
		"       trim(coalesce(current_role, 'NONE'))"
		"  from mon$attachments where mon$attachment_id = current_connection");
	printf("%-22s user=%s auth=%s wirecrypt=%s protocol=%s role=%s\n", label,
		t.rows[0][0].c_str(), t.rows[0][1].c_str(), t.rows[0][2].c_str(),
		t.rows[0][3].c_str(), t.rows[0][4].c_str());
}

static std::string visibleAttachments(fbsample::Db& db, ITransaction* tra)
{
	return db.queryValue(tra,
		"select count(*) from mon$attachments where mon$system_flag = 0");
}

int main()
{
	try
	{
		fbsample::Db admin;
		admin.attachOrCreate(DSN);

		// 1. Layers 1+2, as recorded for THIS attachment.
		{
			ITransaction* t = admin.start();
			whoAmI(admin, t, "admin attachment:");
			t->commit(&admin.status);
		}

		// 2+3. A temporary user (security database) and a privileged role
		//      (this database).  Cleanup first, in case a prior run died.
		//      Every DDL batch gets its own transaction and a FULL commit:
		//      user management is deferred work performed at commit, and on
		//      this Firebird 6 snapshot continuing a transaction after a
		//      COMMIT RETAINING of CREATE USER kills the connection.
		execIgnore(admin, "drop user HANDSON_USER using plugin Srp");
		execIgnore(admin, "drop role HANDSON_MONITOR");
		{
			ITransaction* t = admin.start();
			admin.exec(t, "create user HANDSON_USER password 'Hands0nPw' using plugin Srp");
			t->commit(&admin.status);          // <- the user exists only now
		}
		{
			ITransaction* t = admin.start();
			admin.exec(t,
				"create role HANDSON_MONITOR set system privileges to MONITOR_ANY_ATTACHMENT");
			admin.exec(t, "grant HANDSON_MONITOR to user HANDSON_USER");
			t->commit(&admin.status);
		}

		ITransaction* tra = admin.start();
		printf("\nSEC$USERS (the security database, through the virtual view):\n");
		fbsample::Db::Table users = admin.query(tra,
			"select trim(sec$user_name), trim(sec$plugin), sec$admin"
			"  from sec$users order by 1");
		printf("    %-16s %-8s %s\n", "USER", "PLUGIN", "ADMIN");
		for (const auto& r : users.rows)
			printf("    %-16s %-8s %s\n", r[0].c_str(), r[1].c_str(), r[2].c_str());

		// 3. Same user, without and with the role: the system privilege
		//    decides how much of MON$ is visible.  (admin stays attached,
		//    so a fully-privileged viewer sees at least 2 attachments.)
		printf("\nadmin sees %s user attachments in MON$ATTACHMENTS\n",
			visibleAttachments(admin, tra).c_str());

		{
			fbsample::Db plain;
			plain.attach(DSN, TMP_USER, TMP_PASS);
			ITransaction* t2 = plain.start();
			whoAmI(plain, t2, "user, no role:");
			printf("  -> sees %s attachment(s): only its own\n",
				visibleAttachments(plain, t2).c_str());
			t2->commit(&plain.status);
		}
		{
			fbsample::Db monitor;    // attach WITH the role: DPB + sql_role_name
			IXpbBuilder* dpb = monitor.makeDpb(TMP_USER, TMP_PASS, "UTF8");
			dpb->insertString(&monitor.status, isc_dpb_sql_role_name, "HANDSON_MONITOR");
			monitor.att = monitor.prov->attachDatabase(&monitor.status, DSN,
				dpb->getBufferLength(&monitor.status), dpb->getBuffer(&monitor.status));
			dpb->dispose();
			ITransaction* t3 = monitor.start();
			whoAmI(monitor, t3, "user + role:");
			printf("  -> sees %s attachments: MONITOR_ANY_ATTACHMENT at work\n",
				visibleAttachments(monitor, t3).c_str());
			t3->commit(&monitor.status);
		}

		// 4. The failed login, and its exact error chain.
		printf("\nfailed login (wrong password) produces:\n");
		try
		{
			fbsample::Db bad;
			bad.attach(DSN, TMP_USER, "wrong-password");
		}
		catch (const FbException& e)
		{
			char buf[1024];
			fbsample::master->getUtilInterface()->formatStatus(buf, sizeof(buf), e.getStatus());
			printf("    %s\n", buf);
		}

		tra->commit(&admin.status);

		// Cleanup, again one transaction per DDL batch with a full commit.
		{
			ITransaction* t = admin.start();
			admin.exec(t, "drop user HANDSON_USER using plugin Srp");
			admin.exec(t, "drop role HANDSON_MONITOR");
			t->commit(&admin.status);
		}
		printf("\ntemporary user and role dropped. done.\n");
	}
	catch (const FbException& e)
	{
		return fbsample::report(e);
	}
	return 0;
}

File: samples/rust/src/bin/security.rs
//! security.rs — the security layers observed from Rust.
//!
//! The rsfbclient twin of ../cpp/security.cpp.  In one run:
//!   1. WHO AM I    — the attachment's own MON$ATTACHMENTS row: auth plugin
//!                    and wire-crypt plugin, layers 1 and 2 as the server
//!                    recorded them for THIS connection.
//!   2. USERS       — SEC$USERS, the virtual view over the security database.
//!   3. LEAST PRIVILEGE — a temporary user plus a role carrying the
//!                    MONITOR_ANY_ATTACHMENT system privilege; connecting
//!                    with and without the role (builder .role()) changes how
//!                    much of MON$ATTACHMENTS the same user can see.
//!   4. FAILED LOGIN — the error chain a wrong password produces.
//! User management is DEFERRED work performed at commit — so every CREATE /
//! DROP USER here runs in its own SimpleTransaction with a full commit,
//! exactly the discipline the C++ sample documents.
//! See ../../security-architecture.md.
//!
//! Run (see ../README.md):  cargo run --bin security

use fb_handson_rust::{connect, db_path};
use rsfbclient::{charset, prelude::*, FbError, SimpleConnection, SimpleTransaction};

const TMP_USER: &str = "HANDSON_USER";
const TMP_PASS: &str = "Hands0nPw";

/// Run one statement in its own transaction, ignoring any error — needed for
/// idempotent cleanup, because DROP USER reports "record not found" only at
/// COMMIT (user management is deferred work, executed at commit time).
fn exec_ignore(conn: &mut SimpleConnection, sql: &str) {
    if let Ok(mut t) = SimpleTransaction::new(conn, TransactionConfiguration::default()) {
        if t.execute(sql, ()).is_ok() {
            let _ = t.commit();
        }
    }
}

fn exec_commit(conn: &mut SimpleConnection, sqls: &[&str]) -> Result<(), FbError> {
    let mut t = SimpleTransaction::new(conn, TransactionConfiguration::default())?;
    for sql in sqls {
        t.execute(sql, ())?;
    }
    t.commit() // <- the deferred user-management work happens HERE
}

fn who_am_i(tr: &mut SimpleTransaction, label: &str) -> Result<(), FbError> {
    let (who, auth, wire, proto, role): (
        String,
        Option<String>,
        Option<String>,
        Option<String>,
        String,
    ) = tr
        .query_first(
            "SELECT TRIM(MON$USER), MON$AUTH_METHOD, MON$WIRE_CRYPT_PLUGIN, \
                    MON$REMOTE_PROTOCOL, TRIM(COALESCE(CURRENT_ROLE, 'NONE')) \
             FROM MON$ATTACHMENTS WHERE MON$ATTACHMENT_ID = CURRENT_CONNECTION",
            (),
        )?
        .unwrap();
    let s = |o: Option<String>| o.unwrap_or_else(|| "(none)".into());
    println!(
        "{:<22} user={} auth={} wirecrypt={} protocol={} role={}",
        label, who, s(auth), s(wire), s(proto), role
    );
    Ok(())
}

fn visible_attachments(tr: &mut SimpleTransaction) -> Result<i64, FbError> {
    let (n,): (i64,) = tr
        .query_first(
            "SELECT COUNT(*) FROM MON$ATTACHMENTS WHERE MON$SYSTEM_FLAG = 0",
            (),
        )?
        .unwrap();
    Ok(n)
}

/// Attach to the security scratch database as an arbitrary user, optionally
/// under a role (the builder's .role() is the DPB isc_dpb_sql_role_name).
fn attach_as(u: &str, p: &str, role: Option<&str>) -> Result<SimpleConnection, FbError> {
    let mut b = rsfbclient::builder_native()
        .with_dyn_link()
        .with_remote()
        .host("localhost")
        .db_name(db_path("security"))
        .user(u)
        .pass(p)
        .charset(charset::UTF_8)
        .clone();
    if let Some(r) = role {
        b.role(r);
    }
    Ok(b.connect()?.into())
}

fn main() -> Result<(), FbError> {
    let mut admin = connect("security")?;

    // 1. Layers 1+2, as recorded for THIS attachment.
    {
        let mut t = SimpleTransaction::new(&mut admin, TransactionConfiguration::default())?;
        who_am_i(&mut t, "admin attachment:")?;
        t.commit()?;
    }

    // 2+3. A temporary user (security database) and a privileged role (this
    //      database).  Cleanup first, in case a prior run died.
    exec_ignore(&mut admin, "DROP USER HANDSON_USER USING PLUGIN Srp");
    exec_ignore(&mut admin, "DROP ROLE HANDSON_MONITOR");
    exec_commit(
        &mut admin,
        &["CREATE USER HANDSON_USER PASSWORD 'Hands0nPw' USING PLUGIN Srp"],
    )?;
    exec_commit(
        &mut admin,
        &[
            "CREATE ROLE HANDSON_MONITOR SET SYSTEM PRIVILEGES TO MONITOR_ANY_ATTACHMENT",
            "GRANT HANDSON_MONITOR TO USER HANDSON_USER",
        ],
    )?;

    let mut tra = SimpleTransaction::new(&mut admin, TransactionConfiguration::default())?;
    println!("\nSEC$USERS (the security database, through the virtual view):");
    let users: Vec<(String, String, bool)> = tra.query(
        "SELECT TRIM(SEC$USER_NAME), TRIM(SEC$PLUGIN), SEC$ADMIN FROM SEC$USERS ORDER BY 1",
        (),
    )?;
    println!("    {:<16} {:<8} {}", "USER", "PLUGIN", "ADMIN");
    for (u, p, a) in users {
        println!("    {:<16} {:<8} {}", u, p, a);
    }

    // 3. Same user, without and with the role: the system privilege decides
    //    how much of MON$ is visible.  (admin stays attached, so a
    //    fully-privileged viewer sees at least 2 attachments.)
    println!(
        "\nadmin sees {} user attachments in MON$ATTACHMENTS",
        visible_attachments(&mut tra)?
    );
    tra.commit()?;

    {
        let mut plain = attach_as(TMP_USER, TMP_PASS, None)?;
        let mut t2 = SimpleTransaction::new(&mut plain, TransactionConfiguration::default())?;
        who_am_i(&mut t2, "user, no role:")?;
        println!(
            "  -> sees {} attachment(s): only its own",
            visible_attachments(&mut t2)?
        );
        t2.commit()?;
    }
    {
        let mut monitor = attach_as(TMP_USER, TMP_PASS, Some("HANDSON_MONITOR"))?;
        let mut t3 = SimpleTransaction::new(&mut monitor, TransactionConfiguration::default())?;
        who_am_i(&mut t3, "user + role:")?;
        println!(
            "  -> sees {} attachments: MONITOR_ANY_ATTACHMENT at work",
            visible_attachments(&mut t3)?
        );
        t3.commit()?;
    }

    // 4. The failed login, and its exact error chain.
    println!("\nfailed login (wrong password) produces:");
    match attach_as(TMP_USER, "wrong-password", None) {
        Ok(_) => println!("    BUG: the wrong password was accepted"),
        Err(e) => println!("    {}", e),
    }

    // Cleanup, again one transaction per DDL batch with a full commit.
    exec_commit(
        &mut admin,
        &[
            "DROP USER HANDSON_USER USING PLUGIN Srp",
            "DROP ROLE HANDSON_MONITOR",
        ],
    )?;
    println!("\ntemporary user and role dropped. done.");
    Ok(())
}
